proposals/src/target_system.rs

```diff
@@ -0,0 +1,183 @@
+        scale_codec::Encode,
+        scale_codec::Decode,
+        scale_codec::MaxEncodedLen
+    )
+)]
+pub struct SubstrateTargetSystem {
+    /// Pallet index of proposal handler pallet
+    pallet_index: u8,
+    /// Call index of proposal handler pallet
+    call_index: u8,
+    /// Webb Protocol Merkle `TreeId` (4 bytes).
+    tree_id: u32,
+}
+
+impl SubstrateTargetSystem {
+    /// create new target system for substrate
+    pub fn new(pallet_index: u8, call_index: u8, tree_id: u32) -> Self {
+        Self {
+            pallet_index,
+            call_index,
+            tree_id,
+        }
+    }
+    /// Get pallet index
+    pub fn pallet_index(&self) -> u8 {
+        self.pallet_index
+    }
+    /// Get call index
+    pub fn call_index(&self) -> u8 {
+        self.call_index
+    }
+    /// Get tree_id
+    pub fn tree_id(&self) -> u32 {
+        self.tree_id
+    }
+}
+
+impl TargetSystem {
+    /// Length of the `TargetSystem` (26 bytes).
+    pub const LENGTH: usize = 26;
+    /// Create a new `TargetSystem` as a `ContractAddress`.
+    #[must_use]
+    pub fn new_contract_address<T: Into<[u8; 20]>>(address: T) -> Self {
+        let bytes = address.into();
+        Self::ContractAddress(bytes)
+    }
+    /// Create a new `TargetSystem` for substrate
+    #[must_use]
+    pub fn substrate_target_system(
+        pallet_index: u8,
+        call_index: u8,
+        tree_id: u32,
+    ) -> Self {
+        Self::Substrate(SubstrateTargetSystem::new(
+            pallet_index,
+            call_index,
+            tree_id,
+        ))
+    }
+
+    /// Get the underlying bytes of the `TargetSystem`.
+    #[must_use]
+    pub fn to_bytes(&self) -> [u8; Self::LENGTH] {
+        match self {
+            TargetSystem::ContractAddress(address) => {
+                let mut bytes = [0u8; TargetSystem::LENGTH];
+                let f = 6usize;
+                let t = f + 20;
+                bytes[f..t].copy_from_slice(address);
+                bytes
+            }
+            TargetSystem::Substrate(target_system) => {
+                let mut bytes = [0u8; TargetSystem::LENGTH];
+                let f = 22usize;
+                let t = f + core::mem::size_of::<u32>();
+                bytes[f - 1] = target_system.call_index;
+                bytes[f - 2] = target_system.pallet_index;
+                bytes[f..t]
+                    .copy_from_slice(&target_system.tree_id.to_be_bytes());
+                bytes
+            }
+        }
+    }
+
+    /// Get the underlying bytes of the `TargetSystem` without copying.
+    #[must_use]
+    pub fn into_bytes(self) -> [u8; Self::LENGTH] {
+        self.to_bytes()
+    }
+
+    /// Get susbtrate TargetSystem details
+    #[must_use]
+    pub fn get_substrate_target_system(self) -> SubstrateTargetSystem {
+        match self {
+            TargetSystem::Substrate(target_system) => target_system,
+            _ => SubstrateTargetSystem::new(0, 0, 0),
+        }
+    }
```

Comment:
Why do we need this method? can't we just match on the target system? I do not think it is a good idea to have the API returns a default value when the result is not as expected, this could result in a hard to debug bugs.

If we really need to have a method like this instead of just matching on the variants please return an `Option< SubstrateTargetSystem>` and return `None` when it is not available.